### PR TITLE
API Route: withdraw money

### PR DESCRIPTION
**Pull request by malivinayak**

### CODE REVIEW

File: src/backend/routes/wallet/withdrawMoney.js [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        if (!token || !amount) {
+            return res.status(403).send({
+                status: "failure",
+                message: "Token or amount not received!!!",
+                code: 403,
+            });
+        }
+
+        let connection, query;
+        try {
```
Comment (malivinayak) [inline src/backend/routes/wallet/withdrawMoney.js:42]: missing connection

**Comment by malivinayak** [comment]
@OutdatedGuy API tested using postman

File: src/backend/routes/wallet/withdrawMoney.js [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+if (libPath && fs.existsSync(libPath)) {
+    oracledb.initOracleClient({ libDir: libPath });
+}
+
+const withdrawMoney = async (req, res) => {
+    try {
+        const { token } = req.body;
+        let { amount } = req.body;
+        amount = parseInt(amount);
+        if (amount < 10) {
```
Comment (OutdatedGuy) [inline src/backend/routes/wallet/withdrawMoney.js]: Do 100 instead of 10. Looks standard.

File: src/backend/routes/wallet/withdrawMoney.js [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+            connection = await oracledb.getConnection(dbConfig);
+            options = {
+                outFormat: oracledb.OUT_FORMAT_OBJECT,
+            };
+
+            const getBalQuery = `select WALLET_BALANCE from TRIP_MANAGEMENT_SYSTEM.EMPLOYEE where TOKEN = :1`;
+            const getUserData = await connection.execute(getBalQuery, [token], options);
+            if (getUserData.rows[0] === undefined) {
+                return res.status(403).send({
+                    message: "Invalid Token",
```
Comment (OutdatedGuy) [inline src/backend/routes/wallet/withdrawMoney.js]: Don't send messages like "invalid token" to frontend. The users are not supposed to know this kind of failures. Instead say "Something went wrong, please refresh the page and retry".

File: src/backend/routes/wallet/withdrawMoney.js [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+                    status: "failure",
+                    code: 400,
+                });
+            }
+            let currentBalance = getUserData.rows[0].WALLET_BALANCE;
+            if (amount >= currentBalance) {
+                return res.send({
+                    message: "Insufficient Balance...",
```
Comment (OutdatedGuy) [inline src/backend/routes/wallet/withdrawMoney.js]: Add a little more context/details to the failure message.

File: src/backend/routes/wallet/withdrawMoney.js [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+                    status: "failure",
+                    code: 405,
+                });
+            }
+
+            currentBalance = currentBalance - amount;
+            const updateBalance = `update TRIP_MANAGEMENT_SYSTEM.EMPLOYEE SET WALLET_BALANCE = ${currentBalance} where TOKEN = '${token}'`
+            await connection.execute(updateBalance, [], { autoCommit: true });
+            return res.send({
+                message: `🎉Balance withdrawal successful.......
```
Comment (OutdatedGuy) [inline src/backend/routes/wallet/withdrawMoney.js]: Add space after emoji.

File: src/backend/routes/wallet/withdrawMoney.js [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+To Check updated balance click "Get Wallet Balance" button again`,
+                status: "success",
+                code: 200,
+            });
+        } catch (err) {
+            console.log(" Error at Data Base : " + err);
+            return res.status(500).send({
+                message: "Balance withdraw Failed!!!",
```
Comment (OutdatedGuy) [inline src/backend/routes/wallet/withdrawMoney.js]: Also add "Please try again later" message

**Review by OutdatedGuy — CHANGES_REQUESTED** [review]

File: src/backend/routes/wallet/withdrawMoney.js [hunk 7]
```diff
@@ -49,33 +49,41 @@ const withdrawMoney = async (req, res) => {
...
                 });
             }
+            if (currentBalance < 100) {
+                return res.send({
+                    message: "Insufficient Balance. \nYou must maintain a minimum balance of Rs 100",
+                    status: "failure",
+                    code: 405,
+                });
+
+            }
```
Comment (OutdatedGuy) [inline src/backend/routes/wallet/withdrawMoney.js]: This will never be true? Did you miss something?

**Review by OutdatedGuy — DISMISSED** [review]

File: src/backend/routes/wallet/withdrawMoney.js (hunk 7 above)
Comment (malivinayak) [reply src/backend/routes/wallet/withdrawMoney.js]: Driver has to must maintain min 100Rs balance in his wallet
So after subtracting withdrawal amount if current balance goes below 100 then that transaction will be declined

this must be like
```js
let currentBalance = getUserData.rows[0].WALLET_BALANCE;
currentBalance = currentBalance - amount;
if (currentBalance < 100) {
      return res.send({
          message: "Insufficient Balance. \nYou must maintain a minimum balance of Rs 100",
          status: "failure",
          code: 405,
  });
}
```

Comment (malivinayak) [reply src/backend/routes/wallet/withdrawMoney.js]: Example:
Currently driver has 890Rs in wallet
So he can withdraw less than 790Rs but not more than that (nor less than 100)

image

**Review by OutdatedGuy — APPROVED** [review]

**Pull request merged**